contracts/vesting/src/contract.rs

```diff
@@ -0,0 +1,369 @@
+    _msg: InitMsg,
+) -> Result<Response, ContractError> {
+    Ok(Response::default())
+}
+
+/// Handle an incoming message
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::DelegateToMixnode {
+            mix_identity,
+            delegate_addr,
+            amount,
+        } => try_delegate_to_mixnode(mix_identity, delegate_addr, amount, info, env, deps),
+        ExecuteMsg::UndelegateFromMixnode {
+            mix_identity,
+            delegate_addr,
+        } => try_undelegate_from_mixnode(mix_identity, delegate_addr, info, deps),
+        ExecuteMsg::CreatePeriodicVestingAccount {
+            address,
+            coin,
+            start_time,
+        } => try_create_periodic_vesting_account(address, coin, start_time, info, env, deps),
+        ExecuteMsg::WithdrawVestedCoins { amount } => {
+            try_withdraw_vested_coins(amount, env, info, deps)
+        }
+    }
+}
+
+fn try_withdraw_vested_coins(
+    amount: Coin,
+    env: Env,
+    info: MessageInfo,
+    deps: DepsMut,
+) -> Result<Response, ContractError> {
+    let address = info.sender;
+    if let Some(account) = get_account(deps.storage, &address) {
+        let spendable_coins = account.spendable_coins(None, &env, deps.storage);
+        if amount.amount < spendable_coins.amount {
+            if let Some(balance) = get_account_balance(deps.storage, &address) {
+                let new_balance = balance.u128().saturating_sub(amount.amount.u128());
+                set_account_balance(deps.storage, &address, Uint128(new_balance))?;
+            } else {
+                return Err(ContractError::NoBalanceForAddress(
+                    address.as_str().to_string(),
+                ));
+            }
+
+            let messages = vec![BankMsg::Send {
+                to_address: address.as_str().to_string(),
+                amount: vec![amount],
+            }
+            .into()];
+
+            let attributes = vec![attr("action", "whitdraw")];
+
+            return Ok(Response {
+                submessages: Vec::new(),
+                messages,
+                attributes,
+                data: None,
+            });
+        } else {
+            return Err(ContractError::InsufficientSpendable(
+                address.as_str().to_string(),
+                spendable_coins.amount.u128(),
+            ));
+        }
+    } else {
+        return Err(ContractError::NoAccountForAddress(
+            address.as_str().to_string(),
+        ));
+    }
+    Ok(Response::default())
+}
+
+fn try_delegate_to_mixnode(
+    mix_identity: IdentityKey,
+    delegate_addr: String,
+    amount: Coin,
+    info: MessageInfo,
+    env: Env,
+    deps: DepsMut,
+) -> Result<Response, ContractError> {
+    if info.sender != ADMIN_ADDRESS {
+        return Err(ContractError::NotAdmin(info.sender.as_str().to_string()));
+    }
```

Comment:
I may not be getting the full picture of the story, but I don't really understand why the admin is the one that has to intermediate the delegation as well? Shouldn't it be the vesting account that can do delegations/undelegations, with sender control just like in `try_withdraw_vested_coins`?

Reply:
You're right, at one point I was under the mistaken assumption that the transactions will be coming in from the admin address, and not from the vesting contract address, great catch